presto-main/src/main/java/com/facebook/presto/sql/planner/RowExpressionInterpreter.java

```diff
@@ -781,6 +797,43 @@ private SpecialCallResult tryHandleCast(CallExpression callExpression, List<Obje
                 return changed(null);
             }
 
+            // Only nested casts to varchar types are simplified.
+            if (source instanceof CallExpression && isVarcharType(targetType)) {
+                CallExpression innerCall = (CallExpression) source;
+                Type innerType = innerCall.getType();
+                if (isVarcharType(innerType) && resolution.isCastFunction(innerCall.getFunctionHandle()) && innerCall.getArguments().size() == 1) {
+                    VarcharType narrowestType = narrowVarcharType((VarcharType) targetType, (VarcharType) innerType);
+
+                    RowExpression castArg = innerCall.getArguments().get(0);
+                    if (castArg instanceof VariableReferenceExpression) {
+                        VariableReferenceExpression arg = (VariableReferenceExpression) castArg;
+                        Type argType = arg.getType();
+                        if (isVarcharType(argType)) {
+                            narrowestType = narrowVarcharType(narrowestType, (VarcharType) argType);
+                        }
+                    }
+                    else if (castArg instanceof ConstantExpression) {
+                        ConstantExpression arg = (ConstantExpression) castArg;
+                        Type argType = arg.getType();
+                        if (isVarcharType(argType)) {
+                            narrowestType = narrowVarcharType(narrowestType, (VarcharType) argType);
+                        }
+
+                        FunctionHandle castFunctionHandle = functionManager.lookupCast(CAST, castArg.getType().getTypeSignature(), narrowestType.getTypeSignature());
+                        try {
+                            return changed(functionInvoker.invoke(castFunctionHandle, session.getSqlFunctionProperties(), arg.getValue()));
```

Comment:
Invoking a function is not a good idea especially once we implement user-defined casts etc.

Reply:
I replaced the invoke with a `Slice.slice` call to truncate argument when necessary.